File: src/Samples/CSharpApp/App.ViewModel/GraphViewModel.cs
namespace CSharpApp.ViewModel
{
    using System;
    using System.Windows.Input;
    using Utopia;
    using Utopia.ViewModel;

    public class GraphViewModel : PageViewModel
    {
        ICalculationEngine _calculationEngine;

        public bool[] BooleanTypes { get { return new bool[] { true, false }; } }

        public bool AutoCalculate
        {
            get { return _calculationEngine.Calculation.Automatic; }
            set { _calculationEngine.Calculation.Automatic = value; }
        }

        public GraphViewModel()
        {
            new GraphViewModel(new CalculationEngine());
        }

        public GraphViewModel(ICalculationEngine calculationEngine)
            : base()
        {
            _calculationEngine = calculationEngine;

            Graph = new CustomGraph(_calculationEngine);

            CalculateFullCommand = new ActionCommand(() => Graph.Update("out9"));
            CancelCalculateCommand = new ActionCommand(() => _calculationEngine.Calculation.Cancel());
            CalculatePartialCommand = new ActionCommand(() => Graph.Update("out4"));
            CalculateSecondaryCommand = new ActionCommand(() => Graph.Update("out10"));
        }

        public string LayoutAlgorithmType { get { return "EfficientSugiyama"; } }

        public override string Name { get { return "Graph"; } }

        public CustomGraph Graph { get; private set; }
       
        public ICommand CalculateFullCommand { get; private set; }
        public ICommand CancelCalculateCommand { get; private set; }
        public ICommand CalculatePartialCommand { get; private set; }
        public ICommand CalculateSecondaryCommand { get; private set; }
    }
}

File: src/Samples/FSharpApp/App.ViewModel/GraphViewModel.fs
namespace FSharpApp.ViewModel

open System.Linq
open Utopia
open Utopia.ViewModel

type GraphViewModel(ce : ICalculationEngine) = 
    inherit PageViewModel()
    let graph = new CustomGraph(ce)
    new() = new GraphViewModel(new CalculationEngine())
    member this.BooleanTypes = [| true; false |]
    
    member this.AutoCalculate 
        with get () = ce.Calculation.Automatic
        and set v = ce.Calculation.Automatic <- v
    
    member this.LayoutAlgorithmType = "EfficientSugiyama"
    override this.Name = "Graph"
    member this.CancelCalculateCommand = command <| fun () -> ce.Calculation.Cancel()
    member this.CalculateFullCommand = command <| fun () -> graph.UpdateNode("out9")
    member this.CalculatePartialCommand = command <| fun () -> graph.UpdateNode("out4")
    member this.CalculateSecondaryCommand = command <| fun () -> graph.UpdateNode("out10")
    member this.Graph = graph
